ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/Leaf.java

```diff
@@ -180,6 +188,47 @@ void evaluate() {
         }
     }
 
+    /**
+     * This function computes the allowed number of PSTs for each TSO, as the minimum between the given parameter
+     * and the maximum number of RA reduced by the number of network actions already used
+     */
+    Map<String, Integer> getMaxPstPerTso() {
+        Map<String, Integer> maxPstPerTso = new HashMap<>(treeParameters.getMaxPstPerTso());
+        treeParameters.getMaxRaPerTso().forEach((tso, raLimit) -> {
+            int appliedNetworkActionsForTso = (int) this.networkActions.stream().filter(networkAction -> networkAction.getOperator().equals(tso)).count();
+            int pstLimit =  raLimit - appliedNetworkActionsForTso;
+            maxPstPerTso.put(tso, Math.min(pstLimit, maxPstPerTso.getOrDefault(tso, Integer.MAX_VALUE)));
+        });
+        return maxPstPerTso;
+    }
+
+    /**
+     * This function computes the allowed number of network actions for each TSO, as the minimum between the given
+     * parameter and the maximum number of RA reduced by the number of PSTs already used
+     */
+    Map<String, Integer> getMaxTopoPerTso() {
+        Map<String, Integer> maxTopoPerTso = new HashMap<>(treeParameters.getMaxTopoPerTso());
+        treeParameters.getMaxRaPerTso().forEach((tso, raLimit) -> {
+            int appliedRangeActionsForTso = (int) raoData.getAvailableRangeActions().stream().filter(this::isRangeActionActivated).count();
+            int topoLimit =  raLimit - appliedRangeActionsForTso;
+            maxTopoPerTso.put(tso, Math.min(topoLimit, maxTopoPerTso.getOrDefault(tso, Integer.MAX_VALUE)));
+        });
+        return maxTopoPerTso;
+    }
+
+    boolean isRangeActionActivated(RangeAction rangeAction) {
+        double setpoint = rangeAction.getExtension(RangeActionResultExtension.class).getVariant(raoData.getWorkingVariantId()).getSetPoint(raoData.getOptimizedState().getId());
+        String prePerimeterVariantId = raoData.getCrac().getExtension(ResultVariantManager.class).getPrePerimeterVariantId();
+        double initialSetpoint = rangeAction.getExtension(RangeActionResultExtension.class).getVariant(prePerimeterVariantId).getSetPoint(raoData.getOptimizedState().getId());
+        if (Double.isNaN(setpoint)) {
+            return false;
+        } else if (Double.isNaN(initialSetpoint)) {
+            return true;
```

Comment:
Looks a bit strange
Can this really occur ?

Reply:
it shouldn't

Reply:
(given that the initial situation was copied from the network, but who knows)